unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/DatagramSocketTest.scala

```diff
@@ -0,0 +1,308 @@
+    try {
+      assertEquals(
+        "Expected -1 for remote port as not connected",
+        -1,
+        ds.getPort()
+      );
+      val port = 49152 // any valid port, even if it is unreachable
+      ds.connect(InetAddress.getLocalHost(), port)
+      assertEquals("getPort returned wrong value", port, ds.getPort())
+    } finally {
+      ds.close()
+    }
+  }
+
+  @Test def receiveBufferSize(): Unit = {
+    // This test basically checks that getReceiveBufferSize &
+    // setReceiveBufferSize do not unexpectedly throw and that the former
+    // returns a minimally sane value.
+    //
+    // The Java 8 documentation at URL
+    // https://docs.oracle.com/javase/8/docs/api/java/net/\
+    //     Socket.html#setReceiveBufferSize-int- [sic trailing dash]
+    // describes the argument for setReceiveBufferSize(int) &
+    // setSendBufferSize(int) as a _hint_ to the operating system, _not_
+    // a requirement or demand.  This description is basically unaltered
+    // in Java 10.
+    //
+    // There are a number of reasons the operating system can choose to
+    // ignore the hint. Changing the buffer size, even before a bind() call,
+    // may not be implemented. The buffer size may already be at its
+    // maximum.
+    //
+    // Since, by definition, the OS can ignore the hint, it makes no
+    // sense to set the size, then re-read it and see if it changed.
+    //
+    // The sendBuffersize test refers to this comment.
+    // Please keep both tests synchronized.
+
+    val ds = new DatagramSocket(null)
+    try {
+      val prevValue = ds.getReceiveBufferSize
+      assertTrue(prevValue > 0)
+      ds.setReceiveBufferSize(prevValue + 100)
+    } finally {
+      ds.close()
+    }
+  }
+
+  @Test def sendBufferSize(): Unit = {
+    // This test basically checks that getSendBufferSize &
+    // setSendBufferSize do not unexpectedly throw and that the former
+    // returns a minimally sane value.
+    // See more extensive comments in setBufferSize test.
+
+    val ds = new DatagramSocket(null)
+    try {
+      val prevValue = ds.getSendBufferSize
+      assertTrue(prevValue > 0)
+      ds.setSendBufferSize(prevValue + 100)
+    } finally {
+      ds.close()
+    }
+  }
+
+  @Test def broadcast(): Unit = {
+    val ds = new DatagramSocket(null)
+    try {
+      val prevValue = ds.getBroadcast()
+      assertTrue(prevValue)
+      ds.setBroadcast(!prevValue)
+    } finally {
+      ds.close()
+    }
+  }
+
+  @Test def bind(): Unit = {
+    val ds1 = new DatagramSocket(null)
+    try {
+      val nonLocalAddr =
+        new InetSocketAddress(InetAddress.getByName("123.123.123.123"), 0)
+      assertThrows("ds1", classOf[BindException], ds1.bind(nonLocalAddr))
+    } finally {
+      ds1.close()
+    }
+
+    val ds2 = new DatagramSocket(null)
+    try {
+      ds2.bind(new InetSocketAddress(InetAddress.getLoopbackAddress, 0))
+      val port = ds2.getLocalPort
+      assertTrue("ds2", ds2.isBound())
```

Comment:
Shouldn't this test be moved up on line?  That is,
bind, checkThatBound, getLocalPort?  

I guess that a garbage port from an unbound socket will
not proceed past the eventual isBound() check.  Took me
a few cycles to convince myself that sequence was safe.

Reply:
I reviewed `DatagramSocketTest.scala` until/past_when my eyes crossed.  At this point, I think I will have
to use it awhile.  

I _did_ not check that every method in the JDK API doc had a corresponding test. 

I believe that the SN implementation of `DatagramSocket` will not have the deprecated
`setDatagramSocketImplFactory(DatagramSocketImplFactory fac)` method. I think it
will also not currently be able to implement `getChannel()` (because DatagramChannel is
not yet implemented).

Reply:
I suggest that some future evolution of `DatagramSocketTest.scala` or of a 
hypothetical `Datagram6SocketTest.scala` have at least one test which checks if
IPv6 is enabled and then attempts to send & receive from an IPv6 address.

Currently, I believe that such a send/receive is only done when localhost is IPv6.
One or more  explicit IPv6 tests would increase confidence that IPv6 had been exercised
at least once.